module-services/service-desktop/Sync.cpp

```diff
@@ -0,0 +1,180 @@
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Sync::CreateSyncDir(const std::filesystem::path &path)
+{
+    LOG_INFO("Creating sync directory %s...", path.c_str());
+    if (std::filesystem::exists(path)) {
+        LOG_ERROR("Sync directory already exists!");
+        return false;
+    }
+    std::error_code errorCode;
+    if (!std::filesystem::create_directories(path, errorCode)) {
+        LOG_ERROR("Failed to create directory: %s, error: %d", path.c_str(), errorCode.value());
+        return false;
+    }
+    return true;
+}
+
+bool Sync::PackSyncFiles(const std::filesystem::path &path)
+{
+    if (std::filesystem::is_empty(path)) {
+        LOG_ERROR("Sync dir is empty, quitting...");
+        return false;
+    }
+
+    auto isTarFileOpen                = false;
+    std::filesystem::path tarFilePath = (purefs::dir::getSyncPackagePath() / path.filename());
+    mtar_t tarFile;
+
+    auto cleanup = gsl::finally([&isTarFileOpen, &tarFile]() {
+        if (isTarFileOpen) {
+            mtar_close(&tarFile);
+        }
+    });
+
+    LOG_INFO("Opening tar %s file...", tarFilePath.c_str());
+
+    int ret = mtar_open(&tarFile, tarFilePath.c_str(), "w");
+    if (ret != MTAR_ESUCCESS) {
+        LOG_ERROR("Opening tar file failed, quitting...");
+        return false;
+    }
+    isTarFileOpen = true;
+
+    auto fileStreamBuffer = std::make_unique<char[]>(purefs::buffer::tar_buf);
+    auto tarStreamBuffer  = std::make_unique<char[]>(purefs::buffer::tar_buf);
+    setvbuf(tarFile.stream, tarStreamBuffer.get(), _IOFBF, purefs::buffer::tar_buf);
+    std::error_code errorCode;
+
+    for (auto &direntry : std::filesystem::directory_iterator(path)) {
+        if (!isValidDirentry(direntry)) {
+            continue;
+        }
+
+        LOG_INFO("Archiving file ...");
+        LOG_DEBUG("Writing tar header ...");
+
+        if (mtar_write_file_header(&tarFile,
+                                   direntry.path().filename().c_str(),
+                                   static_cast<unsigned>(std::filesystem::file_size(direntry))) != MTAR_ESUCCESS) {
+            LOG_ERROR("Writing tar header failed");
+            return false;
+        }
+
+        uintmax_t filesize = std::filesystem::file_size(direntry.path(), errorCode);
+        if (errorCode) {
+            LOG_ERROR("Failed to get size for file %s, error: %d", direntry.path().c_str(), errorCode.value());
+            return false;
+        }
+        auto loopCount = (filesize / purefs::buffer::tar_buf) + 1u;
+        uint32_t readSize;
+        std::ifstream fileStream;
+        fileStream.open(direntry.path().string().c_str());
+
+        for (auto i = 0u; i < loopCount; i++) {
+            if (fileStream.bad()) {
+                LOG_ERROR("File stream error.");
+                return false;
+            }
+            if (i + 1u == loopCount) {
+                readSize = filesize % purefs::buffer::tar_buf;
+            }
```

Comment:
So this is the code executed only for the last iteration. Neither `filesize` nor `purefs::buffer::tar_buf` change in this loop. What do you think about not checking this and instead simply read `loopCount-1` blocks of size `tar_buf`and one block of size `filesize % tar_buf`?

Reply:
e.g.:
```
auto readBlock = [&](std::size_t readSize) {
   ...
};
for (auto i = 0u; i < loopCount - 1; i++) {
    readBlock(purefs::buffer::tar_buf);
}
readBlock(filesize % purefs::buffer::tar_buf;
```

Reply:
Your suggestion would work, but then I would have to add two additional checks for values returned from lambdas, making one additional check overall.